### PR TITLE
WIP: prototype for unit support #10349 

### PR DESCRIPTION
**Pull request by Bernhard10**
Trying to follow @jreback idea of adding a dtype for holding the unit.

Related to issue #10349 
- [ ] closes #xxxx
 - [ ] tests added / passed (incomplete tests)
 - [ ] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff`` (except for one warning)
 - [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/dimensioned.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
+import numpy as np
+
+from pandas.core.base import (PandasObject)
+from pandas.util._decorators import cache_readonly
+from pandas import compat
+from pandas.core.common import is_null_slice
+
+
+class Dimensional(PandasObject):
+    """
```
Comment (jreback) [inline pandas/core/dimensioned.py]: Is is meant to be a dimensioned scalar?

Comment (Bernhard10) [inline pandas/core/dimensioned.py]: It is ment to be an array with units. (I did not find a better word for "quantity with unit" than dimensional - see https://english.stackexchange.com/a/48069)

Comment (jreback) [inline pandas/core/dimensioned.py]: units is a property of the dtype, NOT the array itself. You don't need *another* array-like class. Just a proper dtype.

Comment (Bernhard10) [inline pandas/core/dimensioned.py]: I tried that. But I didn't get it to work properly. I guess I don't know enough about the internals of pandas to get it to work.

Comment (Bernhard10) [inline pandas/core/dimensioned.py]: Numpy did not allow me to create an array with my costum dtype. Maybe I missed something in the dtype class construction.

Comment (jreback) [inline pandas/core/dimensioned.py]: we barely use numpy except to hold plain vanilla things. The way to handle this is to simply make a regular float Series with a parameterized dtype of the unit. These would be constructed similar to ``DatetimeTZDtype``, maybe ``float64['ergs']`` or whatever The machinery is already in place to do this.

Comment (Bernhard10) [inline pandas/core/dimensioned.py]: Thanks for the feedback. What you said is exactly what I planned to do, but I somehow don't get it how to do it.

The problem is that `Series` does not seem to habe a dtype other than the numpy array's dtype:
If I understood the code correctly, then the following happens:

If I do `Series([1,2,3], dtype=DimensionedFloatDtype("meter"))`, then the code 
hits `data = _sanitize_array(data, index, dtype, copy, raise_cast_failure=True)` which returns an `np.array`.
This array gets passed into the SingleBlockManager which is passed to `NDFrame.__init__`, where it is stored as `_data`.

Now `Series.dtype` is a property that just retrieves the dtype from the SingleBlockManager (which was constructed with a np.array, and according to `_interleaved_dtype(blocks)` disallows ExtensionDtypes).

The way DatetimeTZDtype does it is by creating a `pandas.core.indexes.DatetimeIndex`, categorical uses `Categorical` to mimic an array.

 I'm sorry if I missed something obvious, but I cannot see how to do this without the need for a wrapper class that holds my float data.

Comment (Bernhard10) [inline pandas/core/dimensioned.py]: The only other option I see would be to explicitly store the dtype in the Series object, but that would be a change that potentially affects things unrelated to numbers with units.

Comment (jreback) [inline pandas/core/dimensioned.py]: you need to step thru code where we use an existing dtype and see how its done. its not easy but all is there. start with simple tests.

Comment (Bernhard10) [inline pandas/core/dimensioned.py]: Sorry, I don't get it.
I looked at `DatetimeTZDtype`, but it has its own block type `DatetimeTZBlock`.
I looked at categorical data, it has its own `class Categorical(PandasObject)`. 
I looked at the PeriodDtype, but it is only used for indexes and I cannot instantiate a Series like `pd.Series([1,2,3], dtype="period[S]")`.

I'm sorry if I overlooked it, but if it is all there, I just don't see where it is.
I can create the Dtype easily enough, I just can't get it to integrate well with pd.Series.

**Comment by Bernhard10** [comment]
If I am allowed to store the dtype in the Series class (for some reason I assumed I was not allowed to do this), then everything is indeed there already in pandas.

The latest commit stored the dtype directly in Series and works straight forward now.
Obviously more tests for edge-cases have to be added.

**Comment by Bernhard10** [comment]
The next step would be to remove the direct dependency on pint and make it generic.

**Comment by Bernhard10** [comment]
With Series storing the dtype, everything else seems to perfectly fall in place.

The current state of the code has (currently) nothing to do with units, but provides a framework how
unit libraries could subclass pandas `ExtensionDtype`. Since `ExtentionDtype` subclasses like DatetimeTZDtype have a lot of special handling all over pandas, I had to subclass `ExtensionType` creating the "abstract" dtype `NumpyDtypeWithMetadata`.

The idea of `NumpyDtypeWithMetadata` is to define an interface for dtypes that store the data in a numpy-compatible dtype but where the dtype holds additional information and defines its own operations.

There is still some work to do:

*) this works only with Series, not with SparseSeries etc.
*) Thus, not all tests pass
*) getting the dtype from a string does not work. It probably will never work, because the actual subclasses of `NumpyDtypeWithMetadata` would be provided by libraries like pint.
*) More thorough testing for edge cases is needed
*) How about indices.


Unfortunately I currently have no time to complete this pull-request, but I will continue to work on it in September, if the basic direction of where this is going is deemed fitting for pandas. Then I could also provide a sample implementation for the use with Pint.

**Comment by jreback** [comment]
@Bernhard10 welcome to have this, though this would require extensive effort / testing, e.g. see for example #16015

**Comment by jreback** [comment]
@Bernhard10 good effort here. going to close. but if you would like to continue working you certainly can. This would need extensive testing in order to be accepted I think.

**Comment by Bernhard10** [comment]
Thanks for the feedback and sorry for the delayed response.

I hope I will find time to continue working on this.

**Pull request closed without merging**